test/Test/Database/LSMTree/Internal/Integration.hs

```diff
@@ -0,0 +1,199 @@
+import           Database.LSMTree.Internal.Serialise
+import           Database.LSMTree.Internal.Serialise.Class
+import           Database.LSMTree.Internal.Serialise.RawBytes as RB
+import           Database.LSMTree.Util
+import           GHC.Generics
+import           Test.Database.LSMTree.Generators (deepseqInvariant,
+                     prop_arbitraryAndShrinkPreserveInvariant)
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans ()
+import           Test.Util.QuickCheck as Util.QC
+
+tests :: TestTree
+tests = testGroup "Test.Database.LSMTree.Internal.Integration" [
+      testGroup "InMemLookupData" $ prop_arbitraryAndShrinkPreserveInvariant (deepseqInvariant @InMemLookupData)
+    , localOption (QuickCheckMaxSize 1000) $
+      testProperty "prop_inMemRunLookupAndConstruction" $ prop_inMemRunLookupAndConstruction
+    ]
+
+-- | Construct a run incrementally, then test a number of positive and negative lookups.
+prop_inMemRunLookupAndConstruction :: InMemLookupData -> Property
+prop_inMemRunLookupAndConstruction dat =
+    Map.size runData > 0 ==>
+        tabulateKeySizes
+      $ tabulateValueSizes
+      $ tabulateNumKeyEntryPairs
+      $ tabulateNumPages
+      $ conjoin (fmap check prepped) .&&. conjoin (fmap check' prepped')
+  where
+    InMemLookupData{runData, lookups} = dat
+
+    tabulateKeySizes = tabulate "Size of key in run" [showPowersOf10 $ sizeofKey k | k <- Map.keys runData ]
+    tabulateValueSizes = tabulate "Size of value in run" [showPowersOf10 $ onValue 0 sizeofValue e | e <- Map.elems runData]
+    tabulateNumKeyEntryPairs = tabulate "Number of key-entry pairs" [showPowersOf10 (Map.size runData) ]
+    tabulateNumPages = tabulate "Number of pages" [showPowersOf10 (Map.size ps) | let (ps, _, _) = run]
+
+    run = mkTestRun runData
+    -- prepLookups says that a key /could be/ in the given page
+    prepped = prepLookups [run] lookups
+    -- prepLookups says that a key /is definitely not/ in the given page
+    prepped' = Set.toList (Set.fromList lookups Set.\\ Set.fromList (fmap fst prepped))
+
+    -- Check that a key /is definitely not/ in the given page.
+    check' :: SerialisedKey -> Property
+    check' k =
+          tabulate1Pre (classifyBin (isJust truth) False)
+        $ tabulate1Post (classifyBin (isJust truth) (isJust test))
+        $ Map.lookup k runData === Nothing
+      where truth = Map.lookup k runData
+            test  = Nothing
+
+    -- | Check that a key /could be/ in the given page
+    check :: (SerialisedKey, (Map Int RawPage, PageSpan)) -> Property
+    check (k, (ps, PageSpan (PageNo i) (PageNo j)))
+      | i <= j    = tabulate "PageSpan distribution" [showPowersOf10 $ j - i + 1]
+                  $ tabulate1Pre (classifyBin (isJust truth) True)
+                  $ tabulate1Post (classifyBin (isJust truth) (isJust test))
+                  $ truth === test
```

Comment:
I like the extensive use of `tabulate`, but I'm not sure how much the "post intra-page lookup" one buys us. Based on the test code here, if the test passes (which I think is the only case we see the tabulation results anyway?), we can only get TruePositive or TrueNegative, where the TruePositive is exactly the same as in the "pre" version. Maybe only having the "pre" version would simplify things a bit?